<h1 align="center">Python performance exercises</h1>

## Python best practices exercises

### Exercise 1

considering the following function for concatenating list strings with delimiter.

In [4]:
def ft_concatenate(l_strings, d):
    """concatenate list of strings into one string separated by delimiter"""
    res = l_strings[0]
    for e in l_strings[1:]:
        res = res + d + e
    return res

- profile the function and identify the bottlenecks.
- improve speed up of the function
*Hint: you may need to look to the string functions in python documentation*

In [5]:
# Utils
import random
import string

def generate_string_list(n, length):
    string_list = []
    for i in range(n):
        characters = string.ascii_letters + string.digits + " "
        ramdom_string = ''.join(random.choice(characters) for _ in range(length))
        string_list.append(ramdom_string)
    return string_list

In [6]:
# data
text = generate_string_list(50000, 10)
delimiter = " * "

In [49]:
# Let profil with timeit
%timeit ft_concatenate(text, delimiter)

706 µs ± 3.81 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [8]:
# Let profil with prun
%prun -s cumulative ft_concatenate(text, delimiter)

         4 function calls in 0.663 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.663    0.663 {built-in method builtins.exec}
        1    0.000    0.000    0.663    0.663 <string>:1(<module>)
        1    0.663    0.663    0.663    0.663 2524129870.py:1(ft_concatenate)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}

In [9]:
!pip install line_profiler
%load_ext line_profiler

In [10]:
%lprun -f ft_concatenate ft_concatenate(text, delimiter)

Timer unit: 1e-09 s

Total time: 0.720666 s
File: /var/folders/w8/b_cflrn97k9c15rcrn5t8mmr0000gn/T/ipykernel_1845/2524129870.py
Function: ft_concatenate at line 1

Line #      Hits         Time  Per Hit   % Time  Line Contents
     1                                           def ft_concatenate(l_strings, d):
     2                                               """concatenate list of strings into one string separated by delimiter"""
     3         1       3000.0   3000.0      0.0      res = l_strings[0]
     4     50000   14949000.0    299.0      2.1      for e in l_strings[1:]:
     5     49999  705709000.0  14114.5     97.9          res = res + d + e
     6         1       5000.0   5000.0      0.0      return res

In [11]:
# !pip install memory_profiler
# %load_ext memory_profiler
# %mprun -f ft_concatenate ft_concatenate(text, delimiter)

In [14]:
# What we have to do is to minimize the number of string concatenations: Here is an optimize version of the function
def ft_concatenate(l_strings, d):
    """concatenate list of strings into one string separated by delimiter"""
    return d.join(l_strings)

In [15]:
# Proof
%lprun -f ft_concatenate ft_concatenate(text, delimiter)

Timer unit: 1e-09 s

Total time: 0.000783 s
File: /var/folders/w8/b_cflrn97k9c15rcrn5t8mmr0000gn/T/ipykernel_1845/4137348578.py
Function: ft_concatenate at line 2

Line #      Hits         Time  Per Hit   % Time  Line Contents
     2                                           def ft_concatenate(l_strings, d):
     3                                               """concatenate list of strings into one string separated by delimiter"""
     4         1     783000.0 783000.0    100.0      return d.join(l_strings)

<div style="padding: .5em; background: #374151; color: #fff; width: 100%">
<h5 style="color:red">Conclusion:</h5>
With this version we could see rthat with just one hit and time=783000. we get the result. And this is much better that oldier result.
</div>

### Exercise 2

In this exercise you will solve the following problem using two methods bruteforce method, and fast method.

**Problem:** You are given a list of n integers, and your task is to calculate the number of distinct values in the list.

**Example**
- Input:
5
2 3 2 2 3

- Output:
2

**Implement the following methods:**

1. **bruteforce method:** create an empty list and start adding items for the given list without adding the previous item add, at the end the result list will contain unique values, print lenght of the list and you are done. 
2. **fast method** think of using Set data structure.

- time the two methods, what do you think?

In [18]:
# bruteforce method
def bruteforce(l):
    distinct_elements = []
    for element in l[1:]:
        if element not in distinct_elements:
            distinct_elements.append(element)
    return len(distinct_elements)

bruteforce([5, 2, 3, 2, 2, 3])


2

In [19]:
# fast method
def fast(l):
    distinct_elements = set(l[1:])
    return len(distinct_elements)

fast([5, 2, 3, 2, 2, 3])

2

In [36]:
import time

In [35]:
# Create a random list of numbers for testing
random_integers = [random.randint(0, 9) for _ in range(10)]
random_integers.insert(0, 10)

# time the two methods
start_time = time.time()
bruteforce(random_integers)
end_time = time.time()
elapsed_time1 = (end_time - start_time)

start_time = time.time()
fast(random_integers)
end_time = time.time()
elapsed_time2 = end_time - start_time

print(f"Bruteforce: \t{elapsed_time1:.6f} seconds\nFast: \t\t{elapsed_time2:.6f} seconds")

Bruteforce: 	0.000028 seconds
Fast: 		0.000026 seconds


## Cython exercises

### Exercise 1

1. load the cython extension.

In [38]:
!pip install cython
%load_ext cython

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 8.9 MB/s eta 0:00:0000:0100:01


2. Considering the following polynomial function:

In [41]:
def poly(a,b):
    return 10.5 * a + 3 * (b**2)

- Create an equivalent Cython function of `poly` with name `poly_cy`.

In [79]:
%%cython -a
def poly_cy(double a, double b):
    return 10.5 * a + 3 * (b**2)

3. time the performance of Python and Cython version of the function, what is the factor of speed up between the two verions.

In [80]:
# write your code here
%timeit poly(34,456)
%timeit poly_cy(34,456)


357 ns ± 0.982 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
80.6 ns ± 0.533 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)


In [81]:
354/80.6

4.392059553349876

<div style="padding: .5em; background: #374151; color: #fff; width: 100%">
<h5 style="color:red">Conclusion:</h5>
The factor of speed up is = 4.39
</div>

4. Now let's work on another example using loop.
    - rewrite the same function below fib that calculates the fibonacci sequence using cython, but now try to add type for the variables used inside it, add a prefix `_cy` to your new cython function.

In [82]:
def fib(n):
    a, b = 1, 1
    for i in range(n):
        a, b = a + b, a

    return a

In [83]:
%%cython -a
def fib_cy(int n):
    cdef int a, b
    a, b = 1, 1
    for i in range(n):
        a, b = a + b, a

    return a


- time the two function for fibonacci series, with n = 20, what is the factor of speed now, What do you think?

In [84]:
%timeit fib(20)
%timeit fib_cy(20)

1.24 µs ± 9.31 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
61.1 ns ± 0.274 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)


In [87]:
factor = 1.24*1000 / 61.1
factor

20.29459901800327

5. Recursive functions are functions that call themselves during their execution. Another interesting property of the Fibonacci sequence is that it can be written as a recursive function. That’s because each item depends on the values of other items (namely item n-1 and item n-2)

- Rewrite the fib function using recursion. Is it faster than the non-recursive version? Does Cythonizing it give even more of an advantage? 

In [17]:
# write your code here


### Exercise 2

- Monte Carlo methods are a broad class of computational algorithms that rely on repeated random sampling to obtain numerical results. 
- One of the basic examples of getting started with the Monte Carlo algorithm is the estimation of Pi.

**Estimation of Pi**

- The idea is to simulate random (x, y) points in a 2-D plane with domain as a square of side 1 unit. 
- Imagine a circle inside the same domain with same diameter and inscribed into the square. 
- We then calculate the ratio of number points that lied inside the circle and total number of generated points. 
- Refer to the image below:

![demo](../data/MonteCarloPlot.png)

We know that area of the square is 1 unit sq while that of circle is $\pi \ast  (\frac{1}{2})^{2} = \frac{\pi}{4}$. Now for a very large number of generated points,

![demo](../data/MonteCarloCalc.png)


## The Algorithm

1. Initialize cile_points, square_points and interval to 0.
2. Generate random point x.
3. Generate random point y.
4. Calculate d = x*x + y*y.
5. If d <= 1, increment circle_points.
6. Increment square_points.
7. Increment interval.
8. If increment < NO_OF_ITERATIONS, repeat from 2.
9. Calculate pi = 4*(circle_points/square_points).
10. Terminate.

**Your mission:** time the function `monte_carlo_pi`, identify the bottlenecks and create a new version using cython functionality to speed up monte carlo simulation for PI, use 100,000 points and compare the speed up factor between python and cython, considering the following optimizations:
- add type for variables used.
- add type for the function
- use c rand function instead of python rand function.
 
*Hint: you can import function from C libraries using the following approach `from libc.<name of c library> cimport <library function name>`, replace the holders `<>` with the right identities for the current problem*

In [18]:
import random
def monte_carlo_pi(nsamples):
    pi = 0.
   # Implement your code here
    return pi

## Numba exercises

### Exercise 1

Previously we considered how to approximateby Monte Carlo.

- Use the same idea here, but make the code efficient using Numba.
- Compare speed with and without Numba when the sample size is large.

In [ ]:
# Your code here

### Exercise 2

In the [Introduction to Quantitative Economics](https://python.quantecon.org/intro.html) with Python lecture series you can learn all about finite-state Markov chains.

For now, let's just concentrate on simulating a very simple example of such a chain.

Suppose that the volatility of returns on an asset can be in one of two regimes — high or low.

The transition probabilities across states are as follows ![markov](../data/markov.png)

For example, let the period length be one day, and suppose the current state is high.

We see from the graph that the state tomorrow will be

- high with probability 0.8

- low with probability 0.2

Your task is to simulate a sequence of daily volatility states according to this rule.

Set the length of the sequence to `n = 1_000_000` and start in the high state.

Implement a pure Python version and a Numba version, and compare speeds.

To test your code, evaluate the fraction of time that the chain spends in the low state.

If your code is correct, it should be about 2/3.

Hints:

- Represent the low state as 0 and the high state as 1.

- If you want to store integers in a NumPy array and then apply JIT compilation, use `x = np.empty(n, dtype=np.int_)`.
